storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -24,149 +26,761 @@ import (
+				"Salt call to GetVolumeSize started",
+			)
+
+			volume.Status.Pending.JobId = jid
+			return self.updateVolumeStatus(volume)
+		}
+	}
+
+	// GetVolumeSize in progress: poll its state.
+	if volume.Status.Pending.Step == storagev1alpha1.GetVolumeSize &&
+		volume.Status.Pending.JobId != "" {
+		jid := volume.Status.Pending.JobId
+		if result, err := self.salt.PollJob(jid, nodeName); err != nil {
+			reqLogger.Error(err, "failed to poll GetVolumeSize status")
+			// This one is not retryable.
+			if failure, ok := err.(*salt.AsyncJobFailed); ok {
+				self.recorder.Event(
+					volume, corev1.EventTypeWarning, "SaltCall",
+					"Salt call to GetVolumeSize failed",
+				)
+
+				return self.setVolumeFailed(
+					volume,
+					storagev1alpha1.CreationError,
+					"GetVolumeSize failed with %s", failure.Error(),
+				)
+			} else {
+				volume.Status.Pending.JobId = ""
+				self.updateVolumeStatus(volume)
+				return reconcile.Result{}, err
+			}
+		} else {
+			if result == nil {
+				reqLogger.Info("GetVolumeSize still in progress")
+				return reconcile.Result{Requeue: true}, nil
+			} else {
+				self.recorder.Event(
+					volume, corev1.EventTypeNormal, "SaltCall",
+					"Salt call to GetVolumeSize succeeded",
+				)
+
+				// TODO(#1461): make this more robust.
+				diskInfo := result["return"].(map[string]interface{})
+				size_str := diskInfo["getsize64"].(string)
+				size_byte, err := strconv.ParseInt(size_str, 10, 64)
+				if err != nil {
+					return self.setVolumeFailed(
+						volume,
+						storagev1alpha1.CreationError,
+						"invalid disk size (%s) from Salt: %s",
+						size_str, err.Error(),
+					)
+				}
+				size := resource.NewQuantity(size_byte, resource.BinarySI)
+				volume.Status.Pending.Step = storagev1alpha1.CreatePV
+				volume.Status.Pending.JobId = ""
+				volume.Status.Pending.Size = *size
+				return self.updateVolumeStatus(volume)
+			}
+		}
+	}
+
+	// Create the PersistentVolume.
+	if volume.Status.Pending.Step == storagev1alpha1.CreatePV {
+		pv := newPersistentVolume(volume, volume.Status.Pending.Size)
+		// Set Volume instance as the owner and controller.
+		err := controllerutil.SetControllerReference(volume, pv, self.scheme)
 		if err != nil {
 			return reconcile.Result{}, err
 		}
+		if err := self.addVolumeFinalizer(volume); err != nil {
+			reqLogger.Error(err, "cannot set volume-protection: requeue")
+			return reconcile.Result{}, err
+		}
+		if err := self.client.Create(context.TODO(), pv); err != nil {
+			reqLogger.Error(
+				err, "cannot create PersistentVolume: requeue",
+				"PersistentVolume.Name", volume.Name,
+			)
+			return reconcile.Result{}, err
+		}
+		reqLogger.Info(
+			"creating a new PersistentVolume", "PersistentVolume.Name", pv.Name,
+		)
+		self.recorder.Event(
+			volume, corev1.EventTypeNormal, "PvCreation",
+			"backing PersistentVolume created",
+		)
+
+		return self.setVolumePhase(volume, storagev1alpha1.VolumeAvailable)
+	}
+
+	// CHECKME
+	panic("unexpected cyka")
```

Comment:
Not sure we should ever `panic`.

Reply:
no we don't want to panic for sure, worst case we go in a `Failed` state.

That's just temporary until I cleanup the code (I just wanted to be sure I wasn't going there in my test).